client.go

```diff
@@ -220,3 +234,110 @@ func (g Client) waitForTx(ctx context.Context, hash string) (*ctypes.ResultTx, e
 
 	return nil, fmt.Errorf("max block wait exceeded")
 }
+
+func NewWSEvents(ctx context.Context) (*WSEvents, error) {
+	wse := &WSEvents{}
+
+	var err error
+	wse.wsc, err = jsonrpcclient.NewWS(viper.GetString("TM_ADDR"),
+		TM_WS_ENDPOINT,
+		jsonrpcclient.PingPeriod(TM_WS_PING_PERIOD),
+		jsonrpcclient.MaxReconnectAttempts(TM_WS_MAX_RECONNECT),
+		jsonrpcclient.OnReconnect(func() {
+			// resubscribe immediately
+			wse.redoSubscriptionsAfter(0 * time.Second)
+		}))
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating ws client")
+	}
+
+	w := logger.FromContext(ctx).WriterLevel(logrus.DebugLevel)
+	l := log.NewTMLogger(log.NewSyncWriter(w))
+	wse.wsc.SetLogger(l)
+
+	if err := wse.wsc.Start(); err != nil {
+		return nil, errors.Wrap(err, "error connecting to WS")
+	}
+
+	return wse, nil
+}
+
+// processes events from tm
+// returns error on failure
+// returns error when event handler returns error
+func (wse *WSEvents) Subscribe(ctx context.Context, query string, h evenHandlerFunc) (<-chan struct{}, chan error) {
+	e := make(chan error)
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		defer cancel()
+		err := wse.wsc.Subscribe(ctx, query)
+		if err != nil {
+			e <- errors.Wrap(err, "error sending subscribe request")
+			return
+		}
+
+		wse.query = query
```

Comment:
good to initialize the entire object. either set query in constructor or dont have this field in object